junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,447 @@
+              val headMethodId = Literal(Constant(id.toString))
+              If(
+                gen.mkMethodCall(paramSym, newTermName("$eq$eq"), Nil,
+                  List(headMethodId)),
+                gen.mkMethodCall(symbol, Nil),
+                mkInvokeJUnitMethodRhs(tail))
+            case Nil =>
+              methodNotFound
+          }
+        }
+        val invokeJUnitMethodRhs =
+          atOwner(invokeJUnitMethodSym)(typer.typed(mkInvokeJUnitMethodRhs(
+            methods.zipWithIndex)))
+        val param = newValDef_compat(paramSym, EmptyTree)()
+        typer.typedDefDef(newDefDef_compat(invokeJUnitMethodSym, invokeJUnitMethodRhs)(
+          vparamss = List(List(param))))
+      }
+
+      def mkGetJUnitMetadataDef(clDef: ClassDef,
+          mlDefOption: Option[ClassDef]): DefDef = {
+        val methods = jUnitAnnotatedMethods(clDef)
+        val mlMethods = mlDefOption.map(jUnitAnnotatedMethods)
+
+        def mkNewInstance[T: TypeTag](params: List[Tree] = Nil): Tree =
+          Apply(Select(New(TypeTree(typeOf[T])), termNames_CONSTRUCTOR), params)
+
+        def liftAnnotations(methodSymbol: Symbol): List[Tree] = {
+          val annotations = methodSymbol.annotations
+          def lift(t: Tree): Tree = {
+            t match {
+              case lit: Literal => lit
+
+              case Select(t2, name) => Select(lift(t2), name)
+
+              case ValDef(mod, name, tpt, rhs) =>
+                ValDef(mod, name, tpt, lift(rhs))
+
+              case _ => t
+            }
+          }
+          def liftList(params: List[Tree]): List[Tree] =
+            params.map(lift)
+
+          annotations.collect {
+            case ann if ann.atp == typeOf[org.junit.Test] =>
+              ann.original match {
+                case Block(stats, _) =>
+                  // TODO: Add support for @Test(timeout=...)
+                  // val newStats = stats
+                  // val newArgs = liftList(ann.args)
+                  // val newAnn = mkNewInstance[org.junit.Test](newArgs)
+                  // Block(newStats, newAnn)
+                  throw new UnsupportedOperationException(
+                    "@Test(timeout = ...) is not yet supported in " +
+                      "Scala.js JUnit Framework")
+
+                case _ => mkNewInstance[org.junit.Test](liftList(ann.args))
+              }
+
+            case ann if ann.atp == typeOf[org.junit.Before] =>
+              mkNewInstance[org.junit.Before]()
+
+            case ann if ann.atp == typeOf[org.junit.After] =>
+              mkNewInstance[org.junit.After]()
+
+            case ann if ann.atp == typeOf[org.junit.BeforeClass] =>
+              mkNewInstance[org.junit.BeforeClass]()
+
+            case ann if ann.atp == typeOf[org.junit.AfterClass] =>
+              mkNewInstance[org.junit.AfterClass]()
+
+            case ann if ann.atp == typeOf[org.junit.Ignore] =>
+              mkNewInstance[org.junit.Ignore](liftList(ann.args))
```

Comment:
Can't you just have a whitelist of type symbols for which you do this? Its always the same...